packages/boxed-expression-component/src/selection/BeeTableSelectionContext.tsx

```diff
@@ -811,37 +827,78 @@ export function BeeTableSelectionContextProvider({ children }: React.PropsWithCh
     const active = selection.active;
 
     const { startRow, endRow, startColumn, endColumn } = getSelectionIterationBoundaries(selection);
+
+    const maxCellCount = currentRefs.get(0)?.size ?? 0;
+
+    Array.from(currentRefs.keys()).forEach((rowIndex) => {
+      for (let c = 0; c < maxCellCount; c++) {
+        currentRefs
+          .get(rowIndex)
+          ?.get(c)
+          ?.forEach((e) => e.setStatus?.({ isActive: false, isEditing: false, isSelected: false }));
+      }
+    });
+
     for (let r = startRow; r <= endRow; r++) {
+      const reindexedUpperRow: Array<number> = new Array<number>();
+      let reIndexUpper = 0;
+
+      if (r + 1 < 0) {
+        for (let colIndex = 0; colIndex < maxCellCount; colIndex++) {
+          if (currentRefs.get(r)?.get(colIndex)) {
+            reindexedUpperRow.push(reIndexUpper);
+            reIndexUpper += getCellColSpan(currentRefs.get(r)?.get(colIndex));
+          }
+        }
+      }
+
       // Select rowIndex cells
       currentRefs
         .get(r)
         ?.get(0)
         ?.forEach((e) => e.setStatus?.({ isActive: false, isEditing: false, isSelected: true }));
 
       for (let c = startColumn; c <= endColumn; c++) {
-        // Select header cells
-        if (startRow >= 0) {
+        if (r < 0) {
           currentRefs
-            .get(-1)
+            .get(r)
             ?.get(c)
             ?.forEach((e) => e.setStatus?.({ isActive: false, isEditing: false, isSelected: true }));
+          if (r + 1 < 0) {
+            const cellCollSpan = getCellColSpan(currentRefs.get(r)?.get(c));
+            for (let colSubIndex = 0; colSubIndex < cellCollSpan; colSubIndex++) {
+              currentRefs
+                .get(r + +1)
+                ?.get(reindexedUpperRow[c] + colSubIndex)
+                ?.forEach((e) => e.setStatus?.({ isActive: false, isEditing: false, isSelected: true }));
+            }
+          }
```

Comment:
If upper row cells was selected that span more columns/cells from lower row, we select whole group in one step.